### PR TITLE
[Codegen][Tuner] Add support for default tuning specs

### PR DESCRIPTION
**Pull request by kuhar**
These default specs are target architecture-specific and will be shipped with the compiler.

* Default specs belong to target plugins and get embedded in `libIREECompiler.so`, just like ukernels.
* Plugins then register their default tuning specs with the default embedded directory.
* We store them as mlir text. We can't easily assemble them as mlir bytecode without taking a circular dependency on iree-opt. We can revisit this in the future and add a new tool `iree-as` that will only link with dialects.
* After the initial loading, we cache the default specs in the IREE codegen dialect transform library manager.
* Add a placeholder spec for gfx942.
* Document and test the inclusion order. User specs come before default specs.

Issue: https://github.com/iree-org/iree/issues/19214

### CODE REVIEW

**Comment by kuhar** [comment]
I forgot to move the remaining tests that rely on the default specs to the rocm plugin directory

File: compiler/src/iree/compiler/Codegen/Common/MaterializeTuningSpecsPass.cpp [hunk 1]
```diff
@@ -79,6 +94,55 @@ static LogicalResult dumpFinalTuningSpecToDir(ModuleOp tuningSpec,
...
+  }
+
+  // Try to look up the default tuning spec for this architecture, if any.
+  StringRef arch = gpuTarget.getArch();
+  std::string defaultTuningSpecName =
+      llvm::formatv("iree_default_tuning_spec_{}.mlir", arch);
+  std::optional<StringRef> defaultTuningSpecSource;
+  EmbeddedDataDirectory::withGlobal([&](EmbeddedDataDirectory &dir) {
+    defaultTuningSpecSource = dir.getFile(defaultTuningSpecName);
+  });
```
Comment (Groverkss) [inline compiler/src/iree/compiler/Codegen/Common/MaterializeTuningSpecsPass.cpp:132]: Can we move these utility functions to get the specs in iree/compiler/Utils or something? This seems fragile and a common place to have these would be nice.

File: compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenDialect.td [hunk 2]
```diff
@@ -36,8 +36,15 @@ def IREECodegen_Dialect : Dialect {
...
+    /// Returns the loaded module for the `libraryPath` transform dialect
+    /// library MLIR input. If the module has already been loaded in the
+    /// past, returns the memoized module without parsing the file again.
+    /// When the `libraryMLIRSource` argument is not empty, uses the
+    /// contents to load the module instead of accessing the filesystem.
+    /// This function is thead-safe.
     FailureOr<::mlir::ModuleOp>
-    getOrLoadTransformLibraryModule(std::string libraryPath);
+    getOrLoadTransformLibraryModule(StringRef libraryPath,
+                                    StringRef libraryMLIRSource = "");
```
Comment (Groverkss) [inline compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenDialect.td]: Can we have two seperate functions doing this instead?

**Review by Groverkss — COMMENTED** [review]
Mostly LGTM, I'm not a bazel / build system expert so could use someone else for reviewing there.

File: compiler/src/iree/compiler/Codegen/Common/MaterializeTuningSpecsPass.cpp (hunk 1 above)
Comment (kuhar) [reply compiler/src/iree/compiler/Codegen/Common/MaterializeTuningSpecsPass.cpp:132]: This is the only place that will look them up, so I don't think moving them elsewhere will make this easier to understand/manage.

**Review by MaheshRavishankar — COMMENTED** [review]
First pass of this looks good to me. Ill probably do one more deeper look into this. But I am happy to stamp and iterate forward. Might be good to wait for Ben Vanik/Benoit Jacob's review.

File: compiler/plugins/target/ROCM/ROCMTarget.cpp [hunk 3]
```diff
@@ -916,11 +912,31 @@ static void addAMDGPUUkernelBitcodeToGlobalEmbeddedDataDirectory() {
...
 }
 
+// Iterate over default tuning spec embedded-data files, and insert them into
+// the EmbeddedDataDirectory singleton.
+static void addAMDGPUDefaultTuningSpecsToGlobalEmbeddedDataDirectory() {
+  EmbeddedDataDirectory::withGlobal([](EmbeddedDataDirectory &dir) {
+    const iree_file_toc_t *toc =
+        iree_default_tuning_specs_amdgpu_bytecode_create();
+    for (size_t i = 0, e = iree_default_tuning_specs_amdgpu_bytecode_size();
+         i != e; ++i) {
```
Comment (bjacob) [inline compiler/plugins/target/ROCM/ROCMTarget.cpp]: Curious, what's the idea of doing this as opposed to

```c++
for (size_t i = 0; i != iree_default_tuning_specs_amdgpu_bytecode_size(); ++i) {
```

File: compiler/plugins/target/ROCM/builtins/tuning/BUILD.bazel [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    "iree_default_tuning_spec_%s.mlir" % gpu_arch
+    for gpu_arch in gpu_archs
+]
+
+iree_c_embed_data(
+    name = "iree_default_tuning_specs_amdgpu_bytecode",
+    srcs = tuning_spec_mlir_files,
+    c_file_output = "iree_default_tuning_specs_amdgpu.c",
+    flatten = True,
+    h_file_output = "iree_default_tuning_specs_amdgpu.h",
```
Comment (bjacob) [inline compiler/plugins/target/ROCM/builtins/tuning/BUILD.bazel]: Is there a specific reason to have `name` different from `{c,h}_file_output` (minus filename extensions)?

**Review by bjacob — APPROVED** [review]

File: compiler/plugins/target/ROCM/ROCMTarget.cpp (hunk 3 above)
Comment (kuhar) [reply compiler/plugins/target/ROCM/ROCMTarget.cpp]: It's to follow the llvm coding standards: https://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop. It's a good default that doesn't require thinking or justification.

File: compiler/src/iree/compiler/Codegen/Common/MaterializeTuningSpecsPass.cpp [hunk 5]
```diff
@@ -101,37 +164,69 @@ struct MaterializeTuningSpecsPass final
...
         ModuleOp::create(loc, "iree_linked_tuning_spec"));
     linkedTuningSpec.get()->setAttr(
         transform::TransformDialect::kWithNamedSequenceAttrName,
         UnitAttr::get(ctx));
-    linkedTuningSpec->insert(linkedTuningSpec->begin(), userTuningSpec.clone());
+    for (auto [idx, spec] : llvm::enumerate(allSpecs)) {
+      ModuleOp clonedSpec = spec.clone();
+      // Make sure there are no symbol name collisions.
+      clonedSpec.setSymName(
+          llvm::formatv("{}_{}", clonedSpec.getSymName().value(), idx).str());
```
Comment (bjacob) [inline compiler/src/iree/compiler/Codegen/Common/MaterializeTuningSpecsPass.cpp:227]: Neat - I didn't realize that `llvm::formatv` allows standard `{}` as opposed to the numbered placeholders like `{0}`, `{1}` that it suggests in its own comments. Should do a pass to align on that -- makes `llvm::formatv` drop-in replaceable by `std::format` / `fmt::format` whenever we need to.

File: compiler/plugins/target/ROCM/builtins/tuning/BUILD.bazel (hunk 4 above)
Comment (kuhar) [reply compiler/plugins/target/ROCM/builtins/tuning/BUILD.bazel]: oh good catch, I forgot to update this after after I removed serialization to mlirbc

File: compiler/src/iree/compiler/Codegen/Common/MaterializeTuningSpecsPass.cpp (hunk 5 above)
Comment (bjacob) [reply compiler/src/iree/compiler/Codegen/Common/MaterializeTuningSpecsPass.cpp:227]: Actually I just realized that `{0}`, `{1}` is also standard. TIL!

**Review by Groverkss — APPROVED** [review]

File: compiler/src/iree/compiler/Codegen/Common/MaterializeTuningSpecsPass.cpp (hunk 5 above)
Comment (bjacob) [reply compiler/src/iree/compiler/Codegen/Common/MaterializeTuningSpecsPass.cpp:227]: https://github.com/iree-org/iree/pull/19426

**Pull request merged**